ui/js/components/auth.js

```diff
@@ -0,0 +1,173 @@
+            a.href = thServiceDomain;
+            this.userCanLogin = a.hostname === $location.host();
+
+            /**
+             * Using a window listener here because I wasn't getting reliable
+             * results from the events from angular-local-storage.
+             */
+            $window.addEventListener("storage", function(e) {
+                if (e.key === "treeherder.user") {
+                    $timeout(function() {
+                        var newUser = JSON.parse(e.newValue);
+                        if (newUser && newUser.email) {
+                            // User was saved to local storage. Use it.
+                            ctrl.setLoggedIn(newUser);
+                        } else {
+                            // This is a tiny hack.  :)
+                            // Right after login, sometimes we get another
+                            // event come in here with a null value for
+                            // ``newValue``.  This happens once every 10 or so
+                            // logins, in my experience.  But localStorage
+                            // is still set to the logged in user.  So we just
+                            // double-check the localStorage value and possibly
+                            // skip the logout.
+                            var storedUser = localStorageService.get("user");
+                            if (!storedUser || !storedUser.loggedin) {
+                                ctrl.setLoggedOut();
+                            }
+                        }
+                    }, 0);
+                }
+            });
+
+            // Ask the back-end if a user is logged in on page load
+            ThUserModel.get().then(function (currentUser) {
+                if (currentUser.email) {
+                    ctrl.setLoggedIn(currentUser);
+                } else {
+                    ctrl.setLoggedOut();
+                }
+            });
+
+            /**
+             * Contact login.taskcluster to log the user in.  Opens a new tab
+             * for the tc-login, which will get closed if it's successful.
+             */
+            ctrl.login = function () {
+                var hash = encodeURIComponent("#");
+                var colon = encodeURIComponent(":");
+                var target = `${$location.protocol()}${colon}//${$location.host()}${colon}${$location.port()}/${hash}/login`;
+                var description = "Treeherder";
+                var url = `https://login.taskcluster.net/?target=${target}&description=${description}`;
+
+                // open a new tab to show the taskcluster auth login page
+                $window.open(url, "_blank");
+            };
+
+            /**
+             * Contact Treeherder back-end to log the user out and invalidate
+             * the session token.  Then updates the UI
+             */
+            ctrl.logout = function () {
+                $http.get(thUrl.getRootUrl("/auth/logout/"));
```

Comment:
If the logout call fails (bad connection etc), then no error message will be shown, and the UI will erroneously show the user as logged out, until next page load.